### PR TITLE
Add a `FlexElement` to support explicit spacers and flex parameters for `Flex` in xilem

### PR DESCRIPTION
**Pull request by Philipp-M**
Takes ideas from #235 (in particular the masonry side).

This implements support for explicit spacers in the `Flex` view and flex parameters for views (see examples `flex` and `mason` for more details).
It also adds a way (`AnyFlexChild`) to dynamically switch between spacers and widgets (with optional flex parameters).
`masonry::widget::Flex::layout` is also updated to be in sync with current druid.

### CODE REVIEW

File: xilem/src/lib.rs
Comment (Philipp-M) [inline xilem/src/lib.rs:145]: Having a builder method for this on the `WidgetViewTrait, is obviously open for discussion, I'm not sure of it yet:

What may or may not make sense from an API standpoint is to support these kinds of methods, when the returned element is also a `WidgetView`, `FlexItem` is not a `WidgetView`, so it can't be used outside of the context of a `Flex` container view.

File: masonry/src/widget/flex.rs
Comment (Philipp-M) [inline masonry/src/widget/flex.rs:381]: I guess that could be removed now?

File: masonry/src/widget/flex.rs
Comment (Philipp-M) [inline masonry/src/widget/flex.rs:596]: I think I remember discussions about this, to avoid crashes (`NaN` and the likes), I set it to a high but not too high number (as vello goes crazy than, with weird floating point artifacts), the issue also exists for the current `layout` method

File: xilem/src/view/flex.rs
Comment (Philipp-M) [inline xilem/src/view/flex.rs:268]: Because of parameterization issues, I needed to add these generic parameters.

I guess we need to do this a lot more for other views as well, to seamlessly support stuff like `Adapt` or in general inference for these variables, when it's not explicit somewhere...

File: masonry/src/widget/flex.rs
Comment (PoignardAzur) [reply masonry/src/widget/flex.rs:596]: > to avoid crashes (NaN and the likes), I set it to a high but not too high number

Pretty sure there's [a Codeless Code fable](http://thecodelesscode.com/case/115) about this. :smiley: 

The fundamental problem is that Masonry's layout model is confused, not that Flex had the wrong maximum constraint. Though I guess this works in the meantime.

**Review by Philipp-M — COMMENTED** [review]
Yeah that's better.

I think in the future it would be great to somehow implement a similar mechanism as `WidgetMut` to access the `Child` to avoid needing access to the parent/flex-container, but I don't think it should be done in this PR, and I'm still not sure if it's worth it...

File: xilem/src/view/flex.rs
Comment (DJMcNab) [inline xilem/src/view/flex.rs:400]: I think we're missing `AnyFlexView` here.

We could also support a `FlexChild` view, which would be:
```rust
enum FlexChild<V>
{
    Child(V, ...),
    Spacer(...)
}
```
although I don't think that has much value if we have `AnyFlexView`.

**Review by DJMcNab — APPROVED** [review]
I haven't carefully examined the Masonry level changes, but I also think that's fine, because we're planning on changing `Flex` to use Taffy at some point anyway.

I think that Masonry flex *needs* to support something akin to CSS `gap`, because currently this significantly regresses all of our examples. That doesn't need to happen in this PR, but it does need to happen extremely soon after this lands.

The Xilem level changes look as I'd expect. If you want to add `AnyFlexView` in this PR, that would be very useful @Philipp-M, but I'm not going to block on it.

I would like to get @PoignardAzur's sign-off on `WidgetMut::reborrow` before landing.

File: xilem/src/view/flex.rs
Comment (Philipp-M) [reply xilem/src/view/flex.rs:400]: You mean, that one could do something like this?:

```rust
fn my_flex_elem() -> FlexChild {...}
```

which would be dynamic, so that it could be for example used in a `Vec`?

Comment (DJMcNab) [reply xilem/src/view/flex.rs:400]: Well, either:
```rust
fn my_flex_elem() -> FlexChild<Box<AnyWidgetView<MyState>>> {...}
```
or
```rust
fn my_flex_elem() -> AnyFlexView<MyState> {...}
```
But essentially that, yes.

**Comment by PoignardAzur** [comment]
Re: `reborrow_mut`, I wonder if you could skip the `is_reborrow` value and just apply merge_up multiple time. It's supposed to be idempotent anyway.

In theory it could have surprising effects if some other code reads the parent state and doesn't expect it to change; in practice the fact that WidgetMut holds a mut reference to said parent state means that will never happen.

**Comment by PoignardAzur** [comment]
Otherwise, LGTM.

**Review by PoignardAzur — APPROVED** [review]

**Comment by DJMcNab** [comment]
> Re: reborrow_mut, I wonder if you could skip the is_reborrow value and just apply merge_up multiple time. It's supposed to be idempotent anyway.

The `is_reborrow` method is entirely an optimisation, because merge_up is a relatively heavy operation

**Comment by PoignardAzur** [comment]
> because merge_up is a relatively heavy operation

It's not that heavy. About a dozen boolean flags getting OR-ed. (That's 16 bytes with bools, 2 with bitblags).

The two non-trivial bits of logic are text registrations and cursor changes, both of which should probably be moved out of merge_up.

The more concerning part is this:

```rust
        self.text_registrations
            .append(&mut child_state.text_registrations);
```

Which we probably don't want to duplicate. Once we remove that line, we can remove the `is_reborrow` flag. In the meantime I'd suggest adding a todo comment, but we can merge either way.

**Comment by Philipp-M** [comment]
> we can remove the `is_reborrow` flag.

Does it hurt otherwise, I mean it's just little bit of extra mental overhead, and I assume `WidgetMut::Drop` will get more expensive in the future.

Btw. I have started with the `AnyFlexView`, it's more complicated than I thought (requiring all the generational id stuff), I want to get that in here as well, as I see value for it too.

And I will likely also add the `FlexExt` stuff, which I agree with

**Comment by PoignardAzur** [comment]
> Does it hurt otherwise, I mean it's just little bit of extra mental overhead,

Fair enough, it's a minor point.

> and I assume `WidgetMut::Drop` will get more expensive in the future.

I don't expect it to get much more expensive, but it's a fair point too.

File: xilem/src/view/flex.rs
Comment (Philipp-M) [inline xilem/src/view/flex.rs:314]: I've tried to do this via a blanket impl, but it doesn't work, since I can't specialize over `View<Element = FlexElement>` and `WidgetView`, and otherwise that impl intersects, so I hardcoded these directly on either `V: WidgetView` as well as for either `FlexItem` and `FlexSpacer`...

File: xilem/examples/mason.rs
Comment (Philipp-M) [inline xilem/examples/mason.rs:48]: I guess this is our catch-all playground example? This also uses `[impl ViewSequence; N]` btw.

**Review by Philipp-M — COMMENTED** [review]
I've started documenting (including doc tests, which at least check if all the typing checks out), added the `WidgetView::boxed` as I think it's an uncontroversal change, and the `FlexExt` trait, and of course added an `AnyFlexChild` view, which in its `impl View` logic is very similar to `impl ViewSequence for Option<impl ViewSequence>`. I've also pulled the `Axis` into `View` and reexported relevant enums for `flex_item` from masonry.

Since this is a bigger update, @DJMcNab I think it would make sense if you check at least the `impl View for AnyFlexChild`.

File: xilem/src/view/flex.rs
Comment (DJMcNab) [reply xilem/src/view/flex.rs:314]: Did you try `View<Element = Pod<W>>` and `View<Element = FlexElement>`? The first should be equivalent to `WidgetView`, but might be more friendly to the compiler

File: xilem/src/view/flex.rs
Comment (DJMcNab) [inline xilem/src/view/flex.rs:333]:
```suggestion
/// Applies [`impl Into<FlexParams>`](`FlexParams`) to the [`View`] `V`, can be used as child of a [`Flex`] View
```

File: xilem/src/view/flex.rs
Comment (Philipp-M) [reply xilem/src/view/flex.rs:314]: Yep I tried, I did not make it to work, unfortunately...

File: xilem/src/view/flex.rs
Comment (DJMcNab) [inline xilem/src/view/flex.rs:349]: Ideally we'd have an image here. We really should set up `linebender.org/assets` at some point

(To be clear, this is fine as-is)

File: xilem/src/view/flex.rs
Comment (DJMcNab) [inline xilem/src/view/flex.rs:607]: We should have a comment at the spacer view that it is an invariant that this view doesn't deal with messages.

File: xilem/src/view/flex.rs
Comment (DJMcNab) [inline xilem/src/view/flex.rs:144]: I think `FlexElement` is only transitory, so it doesn't matter if the spacer case is huge.
That is, we're never storing a `Vec<FlexElement>` for very long.

File: xilem/src/view/flex.rs
Comment (DJMcNab) [reply xilem/src/view/flex.rs:327]: Yeah, this is a choice we have to make, and I think adding them to the View type is the right one, to keep the type inference failures local.

File: xilem/examples/mason.rs
Comment (DJMcNab) [inline xilem/examples/mason.rs:93]:
```suggestion
            flex((flexy_flex_sequence, flex_sequence)).direction(axis),
```

File: xilem/examples/mason.rs
Comment (DJMcNab) [reply xilem/examples/mason.rs:51]: Yeah - it's fine to stick anything in here :laughing:

File: xilem/examples/mason.rs
Comment (DJMcNab) [inline xilem/examples/mason.rs:60]: This feels `FizzBuzz` adjacent - not sure that we should action that though

**Review by DJMcNab — APPROVED** [review]
Thanks! A few minor feedback things, then we can land this.

File: xilem/src/view/flex.rs
Comment (Philipp-M) [reply xilem/src/view/flex.rs:144]: Well actually I had it before without double boxing, this is your code from #235, but I'm also fine without it.

File: xilem/src/view/flex.rs
Comment (Philipp-M) [reply xilem/src/view/flex.rs:349]: Yeah our documentation story can certainly be improved, but I hope this PR does improve it a little bit at least.

File: xilem/examples/mason.rs
Comment (Philipp-M) [reply xilem/examples/mason.rs:60]: Ah yeah, lets make it `FizzBuzz` ^^

File: xilem/src/view/flex.rs
Comment (Philipp-M) [reply xilem/src/view/flex.rs:350]: I also wonder, if the doc-tests could be combined with snapshots, such that this really also tests the logic (while reducing the amount of manual tests) (But I doubt that this is simple, if at all possible within in reason)

File: masonry/src/widget/flex.rs
Comment (DJMcNab) [inline masonry/src/widget/flex.rs:375]: This needs to be removed.

**Pull request merged**
